test/gst-vaapi/transcode/transcoder.py

```diff
@@ -11,45 +11,199 @@
+    },
+    encode = {
+      "avc" : dict(
+        sw = (ALL_PLATFORMS, have_gst_element("x264enc"), ("x264enc", "h264",  "video/x-h264,profile=main", "h264parse")),
+        hw = (AVC_ENCODE_PLATFORMS, have_gst_element("vaapih264enc"), ("vaapih264enc", "h264", "video/x-h264,profile=main", "h264parse", )),
+      ),
+      "hevc-8" : dict(
+        sw = (ALL_PLATFORMS, have_gst_element("x265enc"), ("x265enc", "h265", "video/x-h265,profile=main", "h265parse" )),
+        hw = (HEVC_ENCODE_8BIT_PLATFORMS, have_gst_element("vaapih265enc"), ("vaapih265enc", "h265", "video/x-h265,profile=main", "h265parse")),
+      ),
+      "mpeg2" : dict(
+        sw = (ALL_PLATFORMS, have_gst_element("avenc_mpeg2video"), ("avenc_mpeg2video", "m2v", None, "mpegvideoparse")),
+        hw = (MPEG2_ENCODE_PLATFORMS, have_gst_element("vaapimpeg2enc"), ("vaapimpeg2enc", "m2v", None, "mpegvideoparse")),
+      ),
+      "mjpeg" : dict(
+        sw = (ALL_PLATFORMS, have_gst_element("jpegenc"), ("jpegenc", "mjpeg", None, "jpegparse")),
+        hw = (JPEG_ENCODE_PLATFORMS, have_gst_element("vaapijpegenc"), ("vaapijpegenc", "mjpeg", None, "jpegparse")),
+      ),
+    }
+  )
+
+  # hevc implies hevc 8 bit
+  requirements["encode"]["hevc"] = requirements["encode"]["hevc-8"]
+  requirements["decode"]["hevc"] = requirements["decode"]["hevc-8"]
+
   def before(self):
     self.refctx = []
+  
+  def get_requirements_data(self, ttype, codec, mode):
+      plats, reqs, ffparams = self.requirements[ttype].get(
+        codec, {}).get(mode, ([], (False, False), None))
+      return plats, reqs, ffparams
 
-  def transcode_1to1(self):
-    self.transcoded = get_media()._test_artifact(
-      "{case}_{width}x{height}_{mode}.{dstextension}".format(**vars(self)))
-    get_media().test_call_timeout = vars(self).get("call_timeout", 0)
-    call("gst-launch-1.0 -vf filesrc location={source} ! {gsttrans}"
-      " ! filesink location={transcoded}".format(**vars(self)))
+  def validate_spec(self):
+    from slash.utils.pattern_matching import Matcher
 
-    self.convert_toyuv()
-    self.check_metrics()
+    assert len(self.outputs), "Invalid test case specification, outputs data empty"
+    assert self.mode in ["sw", "hw"], "Invalid test case specification as mode type not valid"
 
-  def convert_toyuv(self):
-    self.decodedYUV = get_media()._test_artifact(
-       "{case}_{width}x{height}_{mode}_{trans_type}.yuv".format(**vars(self)))
-    call("gst-launch-1.0 -vf filesrc location={transcoded}"
-      " ! {gstdecoder2}"
-      " ! videoconvert ! video/x-raw,format=I420"
-      " ! checksumsink2 file-checksum=false qos=false"
-      " frame-checksum=false plane-checksum=false dump-output=true"
-      " dump-location={decodedYUV}".format(**vars(self)))
+    # generate platform list based on test runtime parameters
+    iplats, ireq, _ =  self.get_requirements_data("decode", self.codec, self.mode)
+    platforms = set(iplats)
+    requires = [ireq,]
+
+    for output in self.outputs:
```

Comment:
Assuming gstreamer can only handle one output at a time and we skip all self.outputs > 1, we don't need a loop here.

Reply:
removed loop